### ISSUE TITLE
Access Gradients

### ISSUE THREAD

**Issue by ErnstTmp** [fchollet/keras#3047]
Hi Fellows,
thanks for making Keras so great!

For reinforcement learning, I would need to accumulate the gradients of the neural network with different sets of input parameters. How can I access the gradients in Keras? 

I need two things:

1) I need to aggregate the gradients over different sets of X, I do not want to run full backprop at this stage.

2) In a second stage, I only want to run the backprop with the gradients from 1. I can not do this in the same batch, since 1 and 2 need to run on two different neural networks (of the same topology).

I am using Theano backend, and I am willing to program. Can you give me some hints on where to look? For a first test, I would rather not use GPUs, if this matters.

Thanks and kind regards
Ernst

**Comment by nitish11** [comment]
@ErnstTmp  : I have tried the same using my custom theano function for my own optimizer.

```
import theano

# Convolution Neural Network architecture 
image_model = Sequential()
image_model.add(Convolution2D(8, 3, 3, border_mode='valid', input_shape=(3, 128, 128), init=custom_initialization))
image_model.layers[0].input = x
image_model.add(MaxPooling2D(pool_size=(2, 2)))
image_model.add(Flatten())
image_model.add(Dense(1024, activation='relu', init=custom_initialization))
image_model.build()

sm = Dense(i, activation='softmax', init=custom_initialization)
sm.set_previous(image_model.layers[-1])
sm.build()

params = image_model.params + [for p in sm.params]
grads = theano.gradient.grad(loss, params)
#loss can also be computed using theano function
```

This is just a basic skelton, let me know if it worked or not.

**Comment by ErnstTmp** [comment]
@nitish11:
Thank you very much for your code - I will try it and let you know how it goes!!

**Comment by aravindr93** [comment]
@nitish11  @ErnstTmp  Can you also have a look at issue #3062 for a related question? The solution provided by Nitish doesn't quite work. `image_model.params` doesn't exist for Sequential() class. I wish it would be that simple though :(

**Comment by ErnstTmp** [comment]
@nitish11:
Thank you very much for your code - I will try it and let you know how it goes!!

**Comment by ErnstTmp** [comment]
@aravindr93: thank you very much for directing me to the issue #3062, that was very helpful!

**Comment by nitish11** [comment]
@ErnstTmp :
You can check the method to get the params from the below link. https://github.com/fchollet/keras/blob/cfbaec60c77448e6249f912d4b4d4a8c079f2bcf/keras/engine/topology.py#L841

@aravindr93 : Keras old build has this feature to get the params directly from model.
Check https://github.com/fchollet/keras/blob/af932d34808fa5e6cf8b7d03fbd70b2d1fc52798/keras/models.py#L386

**Comment by ErnstTmp** [comment]
Thanks a lot, very intersting!
Ernst
